File: app/src/main/java/com/rangerdie/runzombiez/mission/MissionEngine.kt
package com.rangerdie.runzombiez.mission

import com.rangerdie.runzombiez.audio.AudioEngine
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/**
 * Drives a [Mission]'s timeline. The MVP mission engine is deterministic and
 * timeline-driven, not GPS-driven (spec section 7) — events fire at fixed
 * offsets from mission start. GPS-triggered events are a future extension
 * that can hook into the same [dispatch] path.
 */
class MissionEngine(private val audioEngine: AudioEngine) {

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
    private var timelineJob: Job? = null
    private var duckRestoreJob: Job? = null

    private val _state = MutableStateFlow(MissionUiState())
    val state: StateFlow<MissionUiState> = _state.asStateFlow()

    fun start(mission: Mission) {
        stopInternal(resetState = false)
        _state.value = MissionUiState(mission = mission, isRunning = true)

        val sortedEvents = mission.timeline.sortedBy { it.atSeconds }
        timelineJob = scope.launch {
            var lastSecond = 0
            for (event in sortedEvents) {
                val waitMs = ((event.atSeconds - lastSecond) * 1000L).coerceAtLeast(0)
                delay(waitMs)
                lastSecond = event.atSeconds
                dispatch(event)
            }
            val remainingMs = ((mission.durationSeconds - lastSecond) * 1000L).coerceAtLeast(0)
            if (remainingMs > 0) delay(remainingMs)
            complete()
        }
    }

    fun stop() = stopInternal(resetState = true)

    private fun stopInternal(resetState: Boolean) {
        timelineJob?.cancel()
        duckRestoreJob?.cancel()
        audioEngine.stopAll()
        if (resetState) _state.value = MissionUiState()
    }

    private fun dispatch(event: MissionEvent) {
        val isDialogue = event.type == MissionEventType.HAVEN_TRANSMISSION ||
            event.type == MissionEventType.SURVIVOR_DIALOGUE
        _state.update {
            it.copy(
                elapsedSeconds = event.atSeconds,
                currentText = event.text ?: it.currentText,
                // Only dialogue events have a speaker — clear it for narration (zombie
                // encounters, story beats) instead of leaving a stale name on screen.
                currentSpeaker = if (isDialogue) event.speaker else null,
                currentArtwork = event.artwork ?: it.currentArtwork
            )
        }

        when (event.type) {
            MissionEventType.MUSIC_CUE ->
                event.audioAsset?.let { audioEngine.play(AudioChannel.MUSIC, it, loop = true, targetVolume = 0.55f) }

            MissionEventType.HAVEN_TRANSMISSION, MissionEventType.SURVIVOR_DIALOGUE -> {
                audioEngine.duck(AudioChannel.MUSIC)
                audioEngine.duck(AudioChannel.AMBIENCE)
                event.audioAsset?.let { audioEngine.play(AudioChannel.VOICE, it, fadeInMs = 150L, targetVolume = 1f) }
                scheduleDuckRestore()
            }

            MissionEventType.ENVIRONMENT_AMBIENCE ->
                event.audioAsset?.let { audioEngine.play(AudioChannel.AMBIENCE, it, loop = true, targetVolume = 0.35f) }

            MissionEventType.ZOMBIE_ENCOUNTER ->
                event.audioAsset?.let {
                    audioEngine.play(AudioChannel.SFX, it, targetVolume = 0.9f, pan = event.pan, fadeInMs = 200L)
                }

            MissionEventType.STORY_BEAT -> Unit // text/artwork already applied above

            MissionEventType.MISSION_COMPLETE -> complete()
        }
    }

    private fun scheduleDuckRestore(afterMs: Long = 4500L) {
        duckRestoreJob?.cancel()
        duckRestoreJob = scope.launch {
            delay(afterMs)
            audioEngine.restore(AudioChannel.MUSIC, to = 0.55f)
            audioEngine.restore(AudioChannel.AMBIENCE, to = 0.35f)
        }
    }

    private fun complete() {
        timelineJob?.cancel()
        _state.update { it.copy(isRunning = false, isComplete = true) }
        audioEngine.stopAll()
    }
}

data class MissionUiState(
    val mission: Mission? = null,
    val isRunning: Boolean = false,
    val isComplete: Boolean = false,
    val elapsedSeconds: Int = 0,
    val currentText: String? = null,
    val currentSpeaker: String? = null,
    val currentArtwork: String? = null
)

File: js/mission-engine.js
// Drives a mission's JSON timeline. Deliberately timeline-driven, not
// GPS-driven, for the MVP (events fire at fixed atSeconds offsets) — GPS can
// hook into the same dispatch() path as a future extension.

export class MissionEngine {
  constructor(audioEngine) {
    this.audioEngine = audioEngine;
    this.listeners = new Set();
    this.state = this.initialState();
    this._running = false;
    this._duckRestoreTimer = null;
  }

  initialState() {
    return {
      mission: null,
      isRunning: false,
      isComplete: false,
      elapsedSeconds: 0,
      currentText: null,
      currentSpeaker: null,
      currentArtwork: null
    };
  }

  onChange(fn) {
    this.listeners.add(fn);
    return () => this.listeners.delete(fn);
  }

  setState(patch) {
    this.state = { ...this.state, ...patch };
    this.listeners.forEach((fn) => fn(this.state));
  }

  async start(mission) {
    this.stop();
    this._running = true;
    this.setState({ ...this.initialState(), mission, isRunning: true });

    const events = [...mission.timeline].sort((a, b) => a.atSeconds - b.atSeconds);
    let lastSecond = 0;
    for (const event of events) {
      if (!this._running) return;
      const waitMs = Math.max(0, (event.atSeconds - lastSecond) * 1000);
      await this.delay(waitMs);
      if (!this._running) return;
      lastSecond = event.atSeconds;
      this.dispatch(event);
    }
    if (!this._running) return;
    const remainingMs = Math.max(0, (mission.durationSeconds - lastSecond) * 1000);
    await this.delay(remainingMs);
    if (!this._running) return;
    this.complete();
  }

  delay(ms) {
    return new Promise((resolve) => {
      this._pendingTimeout = setTimeout(resolve, ms);
    });
  }

  dispatch(event) {
    const isDialogue = event.type === "HAVEN_TRANSMISSION" || event.type === "SURVIVOR_DIALOGUE";
    this.setState({
      elapsedSeconds: event.atSeconds,
      currentText: event.text ?? this.state.currentText,
      // Only dialogue events have a speaker — clear it for narration (zombie
      // encounters, story beats) instead of leaving a stale name on screen.
      currentSpeaker: isDialogue ? event.speaker : null,
      currentArtwork: event.artwork ?? this.state.currentArtwork
    });

    switch (event.type) {
      case "MUSIC_CUE":
        if (event.audioAsset) {
          this.audioEngine.play("MUSIC", event.audioAsset, {
            loop: event.loop ?? true,
            volume: event.volume ?? 0.55
          });
        }
        break;

      case "HAVEN_TRANSMISSION":
      case "SURVIVOR_DIALOGUE":
        this.audioEngine.duck("MUSIC");
        this.audioEngine.duck("AMBIENCE");
        if (event.audioAsset) {
          this.audioEngine.play("VOICE", event.audioAsset, { fadeInMs: 150, volume: 1 });
        }
        this.scheduleDuckRestore();
        break;

      case "ENVIRONMENT_AMBIENCE":
        if (event.audioAsset) {
          this.audioEngine.play("AMBIENCE", event.audioAsset, {
            loop: event.loop ?? true,
            volume: event.volume ?? 0.35
          });
        }
        break;

      case "ZOMBIE_ENCOUNTER":
        if (event.audioAsset) {
          this.audioEngine.play("SFX", event.audioAsset, {
            volume: 0.9,
            pan: event.pan ?? 0,
            fadeInMs: 200
          });
        }
        break;

      case "STORY_BEAT":
        // Text/artwork already applied above; a story beat may also carry a
        // one-shot cue (e.g. a gate creak) on whichever channel it specifies.
        if (event.audioAsset) {
          this.audioEngine.play(event.channel ?? "SFX", event.audioAsset, {
            volume: 0.8,
            pan: event.pan ?? 0,
            fadeInMs: 100
          });
        }
        break;

      case "MISSION_COMPLETE":
        if (event.audioAsset) {
          this.audioEngine.play("SFX", event.audioAsset, { volume: 0.9, fadeInMs: 100 });
        }
        this.complete();
        break;
    }
  }

  scheduleDuckRestore(afterMs = 4500) {
    clearTimeout(this._duckRestoreTimer);
    this._duckRestoreTimer = setTimeout(() => {
      this.audioEngine.restore("MUSIC", 0.55);
      this.audioEngine.restore("AMBIENCE", 0.35);
    }, afterMs);
  }

  complete() {
    this._running = false;
    clearTimeout(this._pendingTimeout);
    clearTimeout(this._duckRestoreTimer);
    this.setState({ isRunning: false, isComplete: true });
    // Stop everything except SFX so a mission-complete sting (if any) can finish playing.
    this.audioEngine.stop("MUSIC");
    this.audioEngine.stop("VOICE");
    this.audioEngine.stop("AMBIENCE");
  }

  stop() {
    this._running = false;
    clearTimeout(this._pendingTimeout);
    clearTimeout(this._duckRestoreTimer);
    this.audioEngine.stopAll();
    this.setState(this.initialState());
  }
}
